File: snuqs/circuit/qreg.py
from __future__ import annotations


class Qreg:
    def __init__(self, name: str, dim: int):
        if dim == 0:
            raise ValueError("Qreg of dim 0 is not allowed")

        self.name = name
        self.dim = dim
        self.qbits = [Qbit(self, i) for i in range(dim)]

    def __getitem__(self, key: int):
        return self.qbits[key]

    def __repr__(self):
        return f"{self.name}"


class Qbit(Qreg):
    def __init__(self, qreg: Qreg, index: int):
        self.qreg = qreg
        self.index = index
        self.dim = 1

    def __repr__(self):
        return f"{self.qreg.name}[{self.index}]"

File: src/circuit/qreg.cc
#include "circuit/qreg.h"

namespace snuqs {

Qreg::Qreg(std::string name, size_t dim) : name_(name), dim_(dim) {
  for (int i = 0; i < dim; ++i) {
    qbits_.push_back(new Qbit(this, i));
  }
}

Qreg::~Qreg() {
  for (auto p : qbits_)
    delete p;
}

Qbit::Qbit(Qreg *qreg, size_t index)
    : Qreg(qreg->name_, 1), qreg_(qreg), index_(index) {}

} // namespace snuqs
